packages/named-locks/src/index.ts

```diff
@@ -159,21 +177,25 @@ export async function doWithLock<T>(
 
 /**
  * Tries to acquire the given lock, executes the provided function with the lock held,
- * and releases the lock once the function has executed. If the lock cannot be acquired,
- * the function is not executed and null is returned.
+ * and releases the lock once the function has executed.
+ *
+ * If the lock cannot be acquired, the function is not executed. If an `onNotAcquired`
+ * function was provided, this function is called and its return value is returned.
+ * Otherwise, `null` is returned.
  */
-export async function tryWithLock<T>(
+export async function tryWithLock<T, U = null>(
   name: string,
-  options: LockOptions,
+  options: WithLockOptions<U>,
   func: () => Promise<T>,
-): Promise<T | null> {
-  const lock = await tryLockAsync(name, options);
-  if (lock == null) return null;
-  try {
-    return await func();
-  } finally {
-    await releaseLockAsync(lock);
-  }
+): Promise<T | U> {
+  return await doWithLock<T, U>(
+    name,
+    {
+      onNotAcquired: () => null as U,
```

Comment:
`onNotAcquired` should eventually be mandatory for this function. I left it alone for now (defaults to returning `null`) to avoid breaking existing usages.